### PR TITLE
Fix broken CI workflow and PR template

### LINKED ISSUE #1040: CI is red because of formatting issues etc (possibly caused by Flutter upgrade)
Somehow related: https://github.com/fzyzcjy/flutter_rust_bridge/pull/1026#issuecomment-1423661229

### PR DESCRIPTION
**Pull request by temeddix**
Close #1040

Currently, when you make a pull request, CI is not passing even if there is no problem. This PR was made to inspect the status of CI to check for any kind of error in the system. Also, PR template has some problems such as lack of information. This PR tries to fix these issues.

## Checklist

- [x] An issue to be fixed by this PR is listed above.
- [x] New tests are added to ensure new features are working. End-to-end tests are usually in the `./frb_example/pure_dart` example, more specifically, `rust/src/api.rs` and `dart/lib/main.dart`.
- [x] The code generator is run and the code is formatted (e.g. via `just refresh_all`).
- [x] If this PR adds/changes features, documentations (in the `./book` folder) are updated.
- [x] CI is passing.

## Remark for PR creator

* New contributors will be blocked by GitHub from running CI, but you can use [a trick](https://github.com/fzyzcjy/flutter_rust_bridge/pull/663#discussion_r962150638) to workaround this, and verify your code using CI by yourself.
* If fzyzcjy does not reply for a few days, maybe he just did not see it, so please ping him.

### CODE REVIEW

**Comment by temeddix** [comment]
### Change logs
- Change `flutter format` into `dart format` in various files as the former is deprecated.
- Remove lines with `sed -i "" -e 's/pub.flutter-io.cn/pub.dartlang.org/g'` from `justfile` as there is no such resource online. This line produces error.
- Run `just refresh_all`

### Change needed
I didn't even know what Justfile was, and probably many others wouldn't too. Also, using this justfile on Windows is hard. I don't know which OS maintainers are using but there is no `bash` on Windows. With powershell, `&&` commands don't work. Of course contributers can use something like `git bash` but I think more explanation would be needed in PR template like this:
> Justfile is a task runner for the command line interface that allows you to run shell commands from a file named `justfile` in your project directory, Tou can use Justfile after [installing it](https://github.com/casey/just). Note that commands written in `justfile` of this repository are expected to be run in `bash`, not `cmd` or `powershell`. Running `just ...` commands in `cmd` or `powershell` will produce erorrs as the syntax is not compatible. You can use `git bash` if you have Git installed on Windows.

**Comment by fzyzcjy** [comment]
Good job!

**Comment by temeddix** [comment]
While fixing, I found out there were bad commits making problems. What I mean by 'bad' is that the result of the commit is not verified properly. Even if the commit is not ready to go into master, they were actually going in directly by some contributers.

For example, take a look at this commit https://github.com/fzyzcjy/flutter_rust_bridge/commit/310228fde5276cf535b3b5dda6db6ca958a613f0. This commit produced some of the core cause of this CI issue, which was so hard to debug. I don't know how this commit got into master, but I think verification process is not sufficient right now.

I think as a repository of this scale, it is time to consider proper verification and review process. By doing that we can avoid issues like below. Always using pull request would be needed IMHO.
image

**Comment by temeddix** [comment]
@fzyzcjy I think the problems are all solved. I wrote down the changes in the previous comment. CI is fully passing on my forked repository, so please review the code :)

**Comment by fzyzcjy** [comment]
> @fzyzcjy I think the problems are all solved. I wrote down the changes in the previous comment. CI is fully passing on my forked repository, so please review the code :)

I will probably review it within a few hours :)

**Comment by temeddix** [comment]
Oh, I thought that all commits are merged into one during the pull. It would be great to do it that way if possible :D. Thanks for the quick reply.

**Comment by temeddix** [comment]
I also want to mention that some of the formats in `yml` and `md` files have changed. This is because of VSCode's auto-formatting feature. I hope to PR formatted codebase after these previous 2 PRs that I have made

**Comment by fzyzcjy** [comment]
> Oh, I thought that all commits are merged into one during the pull. It would be great to do it that way if possible :D. Thanks for the quick reply.

That's "squashing". It is very helpful for large-scale libraries such as `flutter` itself, but I am not very sure whether we need this at our scale (and number of commits per day).

**Comment by fzyzcjy** [comment]
> I also want to mention that some of the formats in yml and md files have changed. This is because of VSCode's auto-formatting feature. I hope to PR uniformly formatted codebase after these previous 2 PRs that I have made, specifically for toml, yml and md files.

Definitely! Feel free to enforce such formatting rules. For example, you may need to add it to (1) github actions (2) justfile

File: justfile [hunk 1]
```diff
@@ -80,38 +80,19 @@ clean:
...
-    (cd frb_example/with_flutter/rust && cargo clippy -- -D warnings)
-    (cd frb_example/pure_dart/dart && dart pub get)
-    (cd frb_example/pure_dart_multi/dart && dart pub get)
-    (cd frb_example/with_flutter && flutter pub get)
-
-    just lint
-
-    sed -i "" -e 's/pub.flutter-io.cn/pub.dartlang.org/g' frb_example/pure_dart/dart/pubspec.lock
-    sed -i "" -e 's/pub.flutter-io.cn/pub.dartlang.org/g' frb_example/pure_dart_multi/dart/pubspec.lock
-    sed -i "" -e 's/pub.flutter-io.cn/pub.dartlang.org/g' frb_example/with_flutter/pubspec.lock
```
Comment (fzyzcjy) [inline justfile:113]: hmm imho `refresh_all` contains some more scripts than `precommit`. So maybe we should add those scripts to `precommit` when we are merging refresh_all into that

**Review by fzyzcjy — COMMENTED** [review]
Well done! Only a few nits

File: .github/workflows/flutter_macos_test.yml [hunk 2]
```diff
@@ -1,27 +1,27 @@
...
 on:
   workflow_call:
     inputs:
       flutter_version:
         required: false
         type: string
-        # TODO should be 3.7.0 but has a weird bug currently https://github.com/fzyzcjy/flutter_rust_bridge/pull/1023#issuecomment-1415852845
-        default: '3.3.10'
+        # TODO should be 3.7.2 but has a weird bug currently https://github.com/fzyzcjy/flutter_rust_bridge/pull/1023#issuecomment-1415852845
+        default: "3.3.10"
```
[the remark this answers is not in the document]
Comment (temeddix) [reply .github/workflows/flutter_macos_test.yml:10]: Oh I think it's gone by using Flutter 3.7.2. I'll erase it out.

File: .github/workflows/ci.yaml [hunk 3]
```diff
@@ -39,7 +41,7 @@ jobs:
 
       - uses: dart-lang/setup-dart@v1
         with:
-          sdk: '2.17.6' # TODO upgrade later #655
+          sdk: "2.17.6" # TODO upgrade later #655
```
[the remark this answers is not in the document]
Comment (temeddix) [reply .github/workflows/ci.yaml:44]: image
Okay, I'll upgrade Dart to 2.19.2

File: justfile (hunk 1 above)
Comment (temeddix) [reply justfile:113]: As I wrote it above that URL is not valid. I think it was pointing to something else before but now it's not. Trying to open that URL leads to 404 error page.
image

Comment (fzyzcjy) [reply justfile:113]: Oh it means: replaces all the `pub.flutter-io.cn` text into `pub.dartlang.org` text

Comment (temeddix) [reply justfile:113]: I get the purpose of that domain, but that `sed -i` line produces an error probably because the resource was removed.

**Comment by temeddix** [comment]
After increasing Dart version to 2.19.2, Valgrind test is complaining...
image

**Comment by fzyzcjy** [comment]
IIRC, we have a "valgrind ignore" file which specifies what should be ignored. Maybe we can add one line there to ignore those 8 bytes, if we are sure those bytes are not a bug (e.g. Dart VM just allocates them deliberately).

File: justfile (hunk 1 above)
Comment (temeddix) [reply justfile:113]: Okay I will

Comment (temeddix) [reply justfile:113]: P.S. I did change the text and did an experiment, but error still remains.
image

Comment (fzyzcjy) [reply justfile:113]: I will check that. btw that is probably b/c I use that on MacOS while you are linux/win?

Comment (temeddix) [reply justfile:113]: Hmm I'm using Windows, I'll try changing the directory slash from / to \\

Comment (temeddix) [reply justfile:113]: image

Comment (temeddix) [reply justfile:113]: I think the web URL is causing the problem

**Comment by temeddix** [comment]
I'm looking at this code:
```python
# valgrind_util.py
...
# Look through the leak details and make sure that we don't have
# any definitely or indirectly lost bytes. We allow possibly lost
# bytes to lower the risk of false positives.
LEAK_RE = r"(?:definitely|indirectly) lost:"
LEAK_LINE_MATCHER = re.compile(LEAK_RE)
LEAK_OKAY_MATCHER = re.compile(r"lost: 0 bytes in 0 blocks")
leaks = []
for line in errors:
    if LEAK_LINE_MATCHER.search(line):
        leaks.append(line)
        if not LEAK_OKAY_MATCHER.search(line):
            # NOTE XXX edited by me
            print('error: valgrind_util find this line is bad:', line)
            # sys.stderr.writelines(errors)
            sys.exit(1)

# Make sure we found the right number of leak lines.
if not len(leaks) in [0, 2, 3]:
    # sys.stderr.writelines(errors) # NOTE XXX edited by me
    print('error: #### Malformed Valgrind output.\n#### Exiting.')
    sys.exit(1)

# Success.
print('valgrind_util thinks it is ok')
sys.exit(0)
```

File: justfile (hunk 1 above)
Comment (temeddix) [reply justfile:113]: > Anyway, just copy those lines and _comment them out_. I will check them later
Okay it's done

**Comment by temeddix** [comment]
@fzyzcjy I did my best to upgrade versions :\ Currently, that macOS test and valgrind test are exceptions of upgrade.

**Comment by fzyzcjy** [comment]
Great!

**Comment by fzyzcjy** [comment]
@all-contributors please add @temeddix for code

**Pull request merged**